File: app/src/main/java/io/github/ovso/righttoknow/vfacilitydetail/VFacilityDetailActivity.java
package io.github.ovso.righttoknow.vfacilitydetail;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;
import butterknife.BindView;
import butterknife.OnClick;
import com.fsn.cauly.CaulyAdInfo;
import com.fsn.cauly.CaulyAdInfoBuilder;
import com.fsn.cauly.CaulyAdView;
import com.fsn.cauly.CaulyAdViewListener;
import io.github.ovso.righttoknow.R;
import io.github.ovso.righttoknow.Security;
import io.github.ovso.righttoknow.main.BaseActivity;
import io.github.ovso.righttoknow.map.MapActivity;

/**
 * Created by jaeho on 2017. 8. 2
 */

public class VFacilityDetailActivity extends BaseActivity implements VFacilityDetailPresenter.View {

  VFacilityDetailPresenter presenter;
  @BindView(R.id.contents_textview) TextView contentsTextView;
  @BindView(R.id.swipe_refresh) SwipeRefreshLayout swipe;

  @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    presenter = new VFacilityDetailPresenterImpl(this);
    presenter.onCreate(savedInstanceState, getIntent());
  }

  @Override public void setTitle(int titleId) {
    super.setTitle(titleId);
  }

  @Override protected int getLayoutResId() {
    return R.layout.activity_vfacilitydetail;
  }

  @Override public void setSupportActionBar() {
    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
  }

  @Override public void showContents(String contents) {
    contentsTextView.setText(contents);
  }

  @Override public void setListener() {
    swipe.setColorSchemeResources(R.color.colorPrimary);
    swipe.setOnRefreshListener(() -> {
      presenter.onRefresh(getIntent());
    });
  }

  @Override public void showAd() {
    CaulyAdView view;
    CaulyAdInfo info =
        new CaulyAdInfoBuilder(Security.CAULY_APP_CODE).effect(CaulyAdInfo.Effect.Circle.toString())
            .build();
    view = new CaulyAdView(this);
    view.setAdInfo(info);
    view.setAdViewListener(new CaulyAdViewListener() {
      @Override public void onReceiveAd(CaulyAdView caulyAdView, boolean b) {

      }

      @Override public void onFailedToReceiveAd(CaulyAdView caulyAdView, int i, String s) {

      }

      @Override public void onShowLandingScreen(CaulyAdView caulyAdView) {

      }

      @Override public void onCloseLandingScreen(CaulyAdView caulyAdView) {

      }
    });

    ViewGroup adContainer = findViewById(R.id.ad_container);
    adContainer.addView(view);
  }

  @Override public void showLoading() {
    swipe.setRefreshing(true);
  }

  @Override public void hideLoading() {
    swipe.setRefreshing(false);
  }

  @Override public void showMessage(int resId) {
    Toast.makeText(this, resId, Toast.LENGTH_SHORT).show();
  }

  @Override public void navigateToMap(String address) {
    Intent intent = new Intent(getApplicationContext(), MapActivity.class);
    intent.putExtra("address", address);
    startActivity(intent);
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    finish();
    return true;
  }

  @OnClick(R.id.share_button) void onShareClick() {
    Intent intent = new Intent();
    intent.setAction(Intent.ACTION_SEND);
    intent.setType("text/plain");
    intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_message));
    intent.putExtra(Intent.EXTRA_TEXT, contentsTextView.getText().toString());
    Intent chooser = Intent.createChooser(intent, getString(R.string.share_message));
    startActivity(chooser);
  }

  @OnClick(R.id.location_button) void onMapClick() {
    presenter.onMapClick(getIntent());
  }

  @Override protected void onDestroy() {
    super.onDestroy();
    presenter.onDestroy();
  }

  @Override public void onBackPressed() {
    super.onBackPressed();
    presenter.onBackPressed();
  }

  @Override protected void onResume() {
    super.onResume();
    if (getIntent().hasExtra("vio_fac_link")) {
      setTitle(R.string.title_vioation_facility_inquiry_detail);
    } else if (getIntent().hasExtra("violator_link")) {
      setTitle(R.string.title_violator_inquiry_detail);
    }
  }

}

File: app/src/main/java/io/github/ovso/righttoknow/ui/vfacilitydetail/VFacilityDetailActivity.kt
package io.github.ovso.righttoknow.ui.vfacilitydetail

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AlertDialog
import io.github.ovso.righttoknow.R
import io.github.ovso.righttoknow.databinding.ActivityVfacilitydetailBinding
import io.github.ovso.righttoknow.exts.viewBinding
import io.github.ovso.righttoknow.framework.AdsActivity
import io.github.ovso.righttoknow.ui.map.MapActivity
import kotlinx.android.synthetic.main.content_vfacilitydetail.*

class VFacilityDetailActivity : AdsActivity(), VFacilityDetailPresenter.View {
  private val presenter by lazy { VFacilityDetailPresenterImpl(this) }

  private val binding by viewBinding(ActivityVfacilitydetailBinding::inflate)

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(binding.root)
    setSupportActionBar(toolbar)
    setTitle(R.string.title_vioation_facility_inquiry_detail)
    presenter.onCreate(savedInstanceState, intent)
    showAd()

    btn_vfac_share.setOnClickListener {
      val intent = Intent()
      intent.action = Intent.ACTION_SEND
      intent.type = "text/plain"
      intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_message))
      intent.putExtra(Intent.EXTRA_TEXT, contents_textview.text.toString())
      val chooser = Intent.createChooser(intent, getString(R.string.share_message))
      startActivity(chooser)
    }
    btn_vfac_location.setOnClickListener {
      presenter.onMapClick(intent)
    }
  }

  override fun setSupportActionBar() {
    if (supportActionBar != null) {
      supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }
  }

  override fun showContents(contents: String) {
    contents_textview.text = contents
  }

  override fun setListener() {
    srl_vfac.setColorSchemeResources(R.color.colorPrimary)
    srl_vfac.setOnRefreshListener { presenter.onRefresh(this@VFacilityDetailActivity.intent) }
  }

  override fun showLoading() {
    srl_vfac.isRefreshing = true
  }

  override fun hideLoading() {
    srl_vfac.isRefreshing = false
  }

  override fun showMessage(resId: Int) {
    AlertDialog.Builder(this)
      .setMessage(resId)
      .setPositiveButton(
        android.R.string.ok
      ) { dialogInterface: DialogInterface, which: Int -> dialogInterface.dismiss() }
      .show()
  }

  override fun navigateToMap(locations: DoubleArray, facName: String) {
    val intent = Intent(this, MapActivity::class.java)
    intent.putExtra("locations", locations)
    intent.putExtra("facName", facName)
    startActivity(intent)
  }

  override fun hideButton() {
    btn_vfac_location.visibility = View.INVISIBLE
    btn_vfac_share.visibility = View.INVISIBLE
  }

  override fun showButton() {
    btn_vfac_location.visibility = View.VISIBLE
    btn_vfac_share.visibility = View.VISIBLE
  }


  override fun onDestroy() {
    super.onDestroy()
    presenter.onDestroy()
  }

  override fun onResume() {
    super.onResume()
    if (intent.hasExtra("vio_fac_link")) {
      setTitle(R.string.title_vioation_facility_inquiry_detail)
    } else if (intent.hasExtra("violator_link")) {
      setTitle(R.string.title_violator_inquiry_detail)
    }
  }
}
